File: src/catalog_forge/parser.py
from __future__ import annotations

import hashlib
import json
import re
from collections import Counter
from decimal import Decimal, InvalidOperation
from typing import Any, Iterable
from urllib.parse import urljoin
from uuid import UUID

from bs4 import BeautifulSoup, Tag

from .models import EvidenceSource, FieldEvidence, ParseOutcome, ProductRecord


_PRICE_RE = re.compile(r"(?P<currency>[$£€₩])?\s*(?P<amount>\d[\d,]*(?:\.\d{1,2})?)")
_CURRENCY_SYMBOLS = {"$": "USD", "£": "GBP", "€": "EUR", "₩": "KRW"}
_REQUIRED_FIELDS = ("title", "price_amount", "currency", "availability")


def _dom_fingerprint(soup: BeautifulSoup) -> str:
    signatures: Counter[str] = Counter()
    for tag in soup.find_all(True):
        classes = ".".join(sorted(tag.get("class", []))[:3])
        signatures[f"{tag.name}.{classes}" if classes else tag.name] += 1
    payload = "|".join(f"{key}:{count}" for key, count in sorted(signatures.items()))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16]


def _iter_json_ld(soup: BeautifulSoup) -> Iterable[dict[str, Any]]:
    for node in soup.select('script[type="application/ld+json"]'):
        try:
            payload = json.loads(node.get_text(strip=True))
        except (json.JSONDecodeError, TypeError):
            continue
        candidates = payload if isinstance(payload, list) else [payload]
        for candidate in candidates:
            if not isinstance(candidate, dict):
                continue
            graph = candidate.get("@graph")
            if isinstance(graph, list):
                candidates.extend(item for item in graph if isinstance(item, dict))
            yield candidate


def _product_json_ld(soup: BeautifulSoup) -> dict[str, Any] | None:
    for candidate in _iter_json_ld(soup):
        raw_type = candidate.get("@type", "")
        types = raw_type if isinstance(raw_type, list) else [raw_type]
        if any(str(value).lower() == "product" for value in types):
            return candidate
    return None


def _parse_price(value: object, currency: str | None = None) -> tuple[Decimal | None, str | None]:
    if value is None:
        return None, currency
    match = _PRICE_RE.search(str(value).replace("\u00a0", " "))
    if not match:
        return None, currency
    try:
        amount = Decimal(match.group("amount").replace(",", ""))
    except InvalidOperation:
        return None, currency
    symbol_currency = _CURRENCY_SYMBOLS.get(match.group("currency") or "")
    return amount, (currency or symbol_currency)


def _text(node: Tag | None) -> str | None:
    return node.get_text(" ", strip=True) if node else None


def _record(
    evidence: dict[str, FieldEvidence],
    name: str,
    value: object,
    source: EvidenceSource,
    confidence: float,
    selector: str | None = None,
) -> None:
    if value in (None, "", []):
        return
    evidence[name] = FieldEvidence(
        value=value,
        source=source,
        confidence=confidence,
        selector=selector,
    )


def parse_product(
    *,
    html: str,
    url: str,
    run_id: UUID,
    target_id: UUID,
    source: str,
) -> ParseOutcome:
    soup = BeautifulSoup(html, "html.parser")
    fingerprint = _dom_fingerprint(soup)
    evidence: dict[str, FieldEvidence] = {}
    warnings: list[str] = []

    json_product = _product_json_ld(soup)
    if json_product:
        offers = json_product.get("offers") or {}
        if isinstance(offers, list):
            offers = offers[0] if offers else {}
        if not isinstance(offers, dict):
            offers = {}
        _record(evidence, "title", json_product.get("name"), EvidenceSource.JSON_LD, 0.99, "$.name")
        _record(evidence, "brand", _brand_name(json_product.get("brand")), EvidenceSource.JSON_LD, 0.98, "$.brand")
        _record(evidence, "category", json_product.get("category"), EvidenceSource.JSON_LD, 0.98, "$.category")
        _record(evidence, "image_url", _first_value(json_product.get("image")), EvidenceSource.JSON_LD, 0.98, "$.image")
        _record(evidence, "external_id", json_product.get("sku") or json_product.get("productID"), EvidenceSource.JSON_LD, 0.99, "$.sku")
        _record(evidence, "availability", _availability(offers.get("availability")), EvidenceSource.JSON_LD, 0.98, "$.offers.availability")
        amount, currency = _parse_price(offers.get("price"), offers.get("priceCurrency"))
        _record(evidence, "price_amount", amount, EvidenceSource.JSON_LD, 0.99, "$.offers.price")
        _record(evidence, "currency", currency, EvidenceSource.JSON_LD, 0.99, "$.offers.priceCurrency")

    if source == "books_to_scrape":
        _parse_books_to_scrape(soup, evidence)

    _generic_fallbacks(soup, url, evidence)
    canonical_url = _canonical_url(soup, url)
    external_id = str(evidence.get("external_id", FieldEvidence(value=canonical_url, source=EvidenceSource.DERIVED, confidence=0.85)).value)
    title = str(evidence["title"].value) if "title" in evidence else ""
    if not title:
        warnings.append("missing_title")
        return ParseOutcome(
            product=None,
            required_field_coverage=_coverage(evidence),
            dom_fingerprint=fingerprint,
            warnings=warnings,
        )

    coverage = _coverage(evidence)
    confidence = sum(item.confidence for item in evidence.values()) / len(evidence)
    for field in _REQUIRED_FIELDS:
        if field not in evidence:
            warnings.append(f"missing_{field}")

    product = ProductRecord(
        run_id=run_id,
        target_id=target_id,
        source=source,
        external_id=external_id,
        canonical_url=canonical_url,
        title=title,
        price_amount=_decimal_value(evidence.get("price_amount")),
        currency=_string_value(evidence.get("currency")),
        availability=_string_value(evidence.get("availability")),
        brand=_string_value(evidence.get("brand")),
        category=_string_value(evidence.get("category")),
        image_url=urljoin(url, _string_value(evidence.get("image_url")) or "") or None,
        confidence=round(confidence, 4),
        dom_fingerprint=fingerprint,
        field_evidence=evidence,
    )
    return ParseOutcome(
        product=product,
        required_field_coverage=coverage,
        dom_fingerprint=fingerprint,
        warnings=warnings,
    )


def discover_product_links(html: str, base_url: str, source: str) -> list[str]:
    soup = BeautifulSoup(html, "html.parser")
    selector = "article.product_pod h3 a" if source == "books_to_scrape" else 'a[data-product-url], a[rel="product"]'
    links: list[str] = []
    for node in soup.select(selector):
        href = node.get("href") or node.get("data-product-url")
        if href:
            absolute = urljoin(base_url, str(href))
            if absolute not in links:
                links.append(absolute)
    return links


def _parse_books_to_scrape(soup: BeautifulSoup, evidence: dict[str, FieldEvidence]) -> None:
    table = {
        _text(row.find("th")): _text(row.find("td"))
        for row in soup.select("table.table-striped tr")
    }
    _record(evidence, "title", _text(soup.select_one(".product_main h1")), EvidenceSource.SOURCE_SELECTOR, 0.97, ".product_main h1")
    _record(evidence, "external_id", table.get("UPC"), EvidenceSource.SOURCE_SELECTOR, 0.99, "table:UPC")
    _record(evidence, "category", table.get("Product Type"), EvidenceSource.SOURCE_SELECTOR, 0.96, "table:Product Type")
    _record(evidence, "availability", table.get("Availability") or _text(soup.select_one(".availability")), EvidenceSource.SOURCE_SELECTOR, 0.96, "table:Availability")
    price_text = table.get("Price (incl. tax)") or _text(soup.select_one(".price_color"))
    amount, currency = _parse_price(price_text)
    _record(evidence, "price_amount", amount, EvidenceSource.SOURCE_SELECTOR, 0.98, "table:Price (incl. tax)")
    _record(evidence, "currency", currency, EvidenceSource.SOURCE_SELECTOR, 0.98, "table:Price (incl. tax)")
    image = soup.select_one(".item.active img")
    _record(evidence, "image_url", image.get("src") if image else None, EvidenceSource.SOURCE_SELECTOR, 0.95, ".item.active img")


def _generic_fallbacks(soup: BeautifulSoup, url: str, evidence: dict[str, FieldEvidence]) -> None:
    if "title" not in evidence:
        node = soup.select_one('meta[property="og:title"]')
        title = node.get("content") if node else _text(soup.select_one("h1"))
        _record(evidence, "title", title, EvidenceSource.SEMANTIC_FALLBACK, 0.86, "og:title|h1")
    if "image_url" not in evidence:
        node = soup.select_one('meta[property="og:image"]')
        _record(evidence, "image_url", node.get("content") if node else None, EvidenceSource.SEMANTIC_FALLBACK, 0.85, "og:image")
    if "price_amount" not in evidence:
        node = soup.select_one('[itemprop="price"], meta[property="product:price:amount"], [data-price]')
        value = node.get("content") or node.get("data-price") or _text(node) if node else None
        currency_node = soup.select_one('[itemprop="priceCurrency"], meta[property="product:price:currency"]')
        currency_value = currency_node.get("content") if currency_node else None
        amount, currency = _parse_price(value, currency_value)
        _record(evidence, "price_amount", amount, EvidenceSource.SEMANTIC_FALLBACK, 0.84, "itemprop=price")
        _record(evidence, "currency", currency, EvidenceSource.SEMANTIC_FALLBACK, 0.84, "itemprop=priceCurrency")
    if "availability" not in evidence:
        node = soup.select_one('[itemprop="availability"], link[itemprop="availability"], [data-availability]')
        value = node.get("content") or node.get("href") or node.get("data-availability") or _text(node) if node else None
        _record(evidence, "availability", _availability(value), EvidenceSource.SEMANTIC_FALLBACK, 0.82, "itemprop=availability")
    evidence.setdefault("external_id", FieldEvidence(value=url, source=EvidenceSource.DERIVED, confidence=0.8, selector="canonical_url"))


def _canonical_url(soup: BeautifulSoup, url: str) -> str:
    node = soup.select_one('link[rel="canonical"]')
    return urljoin(url, str(node.get("href"))) if node and node.get("href") else url


def _brand_name(value: object) -> object:
    if isinstance(value, dict):
        return value.get("name")
    return value


def _first_value(value: object) -> object:
    if isinstance(value, list):
        return value[0] if value else None
    return value


def _availability(value: object) -> str | None:
    if value is None:
        return None
    return str(value).rsplit("/", 1)[-1].strip()


def _coverage(evidence: dict[str, FieldEvidence]) -> float:
    return sum(field in evidence for field in _REQUIRED_FIELDS) / len(_REQUIRED_FIELDS)


def _string_value(field: FieldEvidence | None) -> str | None:
    return str(field.value) if field and field.value is not None else None


def _decimal_value(field: FieldEvidence | None) -> Decimal | None:
    if not field or field.value is None:
        return None
    return Decimal(str(field.value))

File: web/app/playground/parser.ts
export type PreviewSource = "generic" | "books_to_scrape";

export type PreviewEvidence = {
  value: string;
  source: "json_ld" | "source_selector" | "semantic_fallback" | "derived";
  confidence: number;
  selector: string;
};

export type PreviewProduct = {
  mode: "preview_not_persisted";
  source: PreviewSource;
  external_id: string;
  canonical_url: string;
  title: string;
  price_amount: string | null;
  currency: string | null;
  availability: string | null;
  brand: string | null;
  category: string | null;
  image_url: string | null;
  confidence: number;
  required_field_coverage: number;
  dom_fingerprint: string;
  extraction_path: string[];
  review_required: boolean;
  review_reason: string | null;
};

export type PreviewResult = {
  product: PreviewProduct | null;
  evidence: Record<string, PreviewEvidence>;
  required_field_coverage: number;
  dom_fingerprint: string;
  warnings: string[];
  decision: {
    status: "eligible" | "review" | "rejected";
    label: string;
    reason: string;
  };
};

const REQUIRED_FIELDS = ["title", "price_amount", "currency", "availability"] as const;
const CURRENCY_SYMBOLS: Record<string, string> = { "$": "USD", "£": "GBP", "€": "EUR", "₩": "KRW" };
const PRICE_PATTERN = /([$£€₩])?\s*(\d[\d,]*(?:\.\d{1,2})?)/;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function record(
  evidence: Record<string, PreviewEvidence>,
  field: string,
  rawValue: unknown,
  source: PreviewEvidence["source"],
  confidence: number,
  selector: string,
) {
  if (rawValue === null || rawValue === undefined || rawValue === "" || (Array.isArray(rawValue) && rawValue.length === 0)) return;
  evidence[field] = { value: String(rawValue), source, confidence, selector };
}

function text(node: Element | null) {
  const value = node?.textContent?.replace(/\s+/g, " ").trim();
  return value || null;
}

function attribute(node: Element | null, name: string) {
  const value = node?.getAttribute(name)?.trim();
  return value || null;
}

function parsePrice(value: unknown, currency?: string | null) {
  if (value === null || value === undefined) return { amount: null, currency: currency || null };
  const match = String(value).replace(/\u00a0/g, " ").match(PRICE_PATTERN);
  if (!match) return { amount: null, currency: currency || null };
  return {
    amount: match[2].replace(/,/g, ""),
    currency: currency || CURRENCY_SYMBOLS[match[1] || ""] || null,
  };
}

function availability(value: unknown) {
  if (value === null || value === undefined) return null;
  return String(value).split("/").at(-1)?.trim() || null;
}

function absoluteUrl(value: string | null, baseUrl: string) {
  if (!value) return baseUrl;
  try {
    return new URL(value, baseUrl).toString();
  } catch {
    return value;
  }
}

function brandName(value: unknown) {
  return isRecord(value) ? value.name : value;
}

function firstValue(value: unknown) {
  return Array.isArray(value) ? value[0] : value;
}

function productJsonLd(document: Document) {
  const candidates: unknown[] = [];
  for (const node of document.querySelectorAll('script[type="application/ld+json"]')) {
    try {
      const parsed: unknown = JSON.parse(node.textContent?.trim() || "null");
      candidates.push(...(Array.isArray(parsed) ? parsed : [parsed]));
    } catch {
      continue;
    }
  }
  for (let index = 0; index < candidates.length; index += 1) {
    const candidate = candidates[index];
    if (!isRecord(candidate)) continue;
    if (Array.isArray(candidate["@graph"])) candidates.push(...candidate["@graph"]);
    const rawType = candidate["@type"];
    const types = Array.isArray(rawType) ? rawType : [rawType];
    if (types.some(value => String(value).toLowerCase() === "product")) return candidate;
  }
  return null;
}

function parseJsonLd(document: Document, evidence: Record<string, PreviewEvidence>) {
  const product = productJsonLd(document);
  if (!product) return;
  const rawOffers = product.offers;
  const offersCandidate = Array.isArray(rawOffers) ? rawOffers[0] : rawOffers;
  const offers = isRecord(offersCandidate) ? offersCandidate : {};
  record(evidence, "title", product.name, "json_ld", .99, "$.name");
  record(evidence, "brand", brandName(product.brand), "json_ld", .98, "$.brand");
  record(evidence, "category", product.category, "json_ld", .98, "$.category");
  record(evidence, "image_url", firstValue(product.image), "json_ld", .98, "$.image");
  record(evidence, "external_id", product.sku || product.productID, "json_ld", .99, "$.sku");
  record(evidence, "availability", availability(offers.availability), "json_ld", .98, "$.offers.availability");
  const price = parsePrice(offers.price, offers.priceCurrency ? String(offers.priceCurrency) : null);
  record(evidence, "price_amount", price.amount, "json_ld", .99, "$.offers.price");
  record(evidence, "currency", price.currency, "json_ld", .99, "$.offers.priceCurrency");
}

function parseBooksSelector(document: Document, evidence: Record<string, PreviewEvidence>) {
  const table = new Map<string, string>();
  for (const row of document.querySelectorAll("table.table-striped tr")) {
    const key = text(row.querySelector("th"));
    const value = text(row.querySelector("td"));
    if (key && value) table.set(key, value);
  }
  record(evidence, "title", text(document.querySelector(".product_main h1")), "source_selector", .97, ".product_main h1");
  record(evidence, "external_id", table.get("UPC"), "source_selector", .99, "table:UPC");
  record(evidence, "category", table.get("Product Type"), "source_selector", .96, "table:Product Type");
  record(evidence, "availability", table.get("Availability") || text(document.querySelector(".availability")), "source_selector", .96, "table:Availability");
  const price = parsePrice(table.get("Price (incl. tax)") || text(document.querySelector(".price_color")));
  record(evidence, "price_amount", price.amount, "source_selector", .98, "table:Price (incl. tax)");
  record(evidence, "currency", price.currency, "source_selector", .98, "table:Price (incl. tax)");
  record(evidence, "image_url", attribute(document.querySelector(".item.active img"), "src"), "source_selector", .95, ".item.active img");
}

function parseSemanticFallback(document: Document, pageUrl: string, evidence: Record<string, PreviewEvidence>) {
  if (!evidence.title) {
    const openGraph = document.querySelector('meta[property="og:title"]');
    record(evidence, "title", attribute(openGraph, "content") || text(document.querySelector("h1")), "semantic_fallback", .86, "og:title|h1");
  }
  if (!evidence.image_url) {
    record(evidence, "image_url", attribute(document.querySelector('meta[property="og:image"]'), "content"), "semantic_fallback", .85, "og:image");
  }
  if (!evidence.price_amount) {
    const priceNode = document.querySelector('[itemprop="price"], meta[property="product:price:amount"], [data-price]');
    const priceValue = attribute(priceNode, "content") || attribute(priceNode, "data-price") || text(priceNode);
    const currencyNode = document.querySelector('[itemprop="priceCurrency"], meta[property="product:price:currency"]');
    const price = parsePrice(priceValue, attribute(currencyNode, "content"));
    record(evidence, "price_amount", price.amount, "semantic_fallback", .84, "itemprop=price");
    record(evidence, "currency", price.currency, "semantic_fallback", .84, "itemprop=priceCurrency");
  }
  if (!evidence.availability) {
    const node = document.querySelector('[itemprop="availability"], link[itemprop="availability"], [data-availability]');
    const value = attribute(node, "content") || attribute(node, "href") || attribute(node, "data-availability") || text(node);
    record(evidence, "availability", availability(value), "semantic_fallback", .82, "itemprop=availability");
  }
  if (!evidence.external_id) record(evidence, "external_id", pageUrl, "derived", .8, "canonical_url");
}

async function fingerprint(document: Document) {
  const counts = new Map<string, number>();
  for (const element of document.querySelectorAll("*")) {
    const classes = Array.from(element.classList).sort().slice(0, 3).join(".");
    const signature = classes ? `${element.tagName.toLowerCase()}.${classes}` : element.tagName.toLowerCase();
    counts.set(signature, (counts.get(signature) || 0) + 1);
  }
  const payload = Array.from(counts.entries()).sort(([left], [right]) => left.localeCompare(right)).map(([key, count]) => `${key}:${count}`).join("|");
  if (globalThis.crypto?.subtle) {
    const digest = await globalThis.crypto.subtle.digest("SHA-256", new TextEncoder().encode(payload));
    return Array.from(new Uint8Array(digest)).map(value => value.toString(16).padStart(2, "0")).join("").slice(0, 16);
  }
  let fallback = 2166136261;
  for (const character of payload) fallback = Math.imul(fallback ^ character.charCodeAt(0), 16777619);
  return Math.abs(fallback >>> 0).toString(16).padStart(16, "0").slice(0, 16);
}

export function validatePreviewUrl(value: string) {
  try {
    const parsed = new URL(value);
    return parsed.protocol === "http:" || parsed.protocol === "https:" ? null : "http 또는 https URL만 입력할 수 있습니다.";
  } catch {
    return "올바른 절대 URL을 입력해 주세요.";
  }
}

export async function parseProductPreview(html: string, pageUrl: string, source: PreviewSource): Promise<PreviewResult> {
  const document = new DOMParser().parseFromString(html, "text/html");
  const evidence: Record<string, PreviewEvidence> = {};
  parseJsonLd(document, evidence);
  if (source === "books_to_scrape") parseBooksSelector(document, evidence);
  parseSemanticFallback(document, pageUrl, evidence);

  const domFingerprint = await fingerprint(document);
  const coverage = REQUIRED_FIELDS.filter(field => evidence[field]).length / REQUIRED_FIELDS.length;
  const warnings = REQUIRED_FIELDS.filter(field => !evidence[field]).map(field => `missing_${field}`);
  const canonicalNode = document.querySelector('link[rel="canonical"]');
  const canonicalUrl = absoluteUrl(attribute(canonicalNode, "href"), pageUrl);
  const title = evidence.title?.value || "";

  if (!title) {
    return {
      product: null,
      evidence,
      required_field_coverage: coverage,
      dom_fingerprint: domFingerprint,
      warnings: ["missing_title", ...warnings.filter(value => value !== "missing_title")],
      decision: { status: "rejected", label: "상품으로 인식하지 못함", reason: "제목을 찾을 수 없어 product record를 만들지 않았습니다." },
    };
  }

  const confidence = Object.values(evidence).reduce((sum, item) => sum + item.confidence, 0) / Object.keys(evidence).length;
  const reviewReason = coverage < 1 ? "incomplete_required_fields" : confidence < .9 ? "low_confidence" : null;
  const extractionPath = Array.from(new Set(REQUIRED_FIELDS.map(field => evidence[field]?.source).filter((value): value is PreviewEvidence["source"] => Boolean(value))));
  const product: PreviewProduct = {
    mode: "preview_not_persisted",
    source,
    external_id: evidence.external_id?.value || canonicalUrl,
    canonical_url: canonicalUrl,
    title,
    price_amount: evidence.price_amount?.value || null,
    currency: evidence.currency?.value || null,
    availability: evidence.availability?.value || null,
    brand: evidence.brand?.value || null,
    category: evidence.category?.value || null,
    image_url: absoluteUrl(evidence.image_url?.value || null, pageUrl) || null,
    confidence: Math.round(confidence * 10_000) / 10_000,
    required_field_coverage: coverage,
    dom_fingerprint: domFingerprint,
    extraction_path: extractionPath,
    review_required: reviewReason !== null,
    review_reason: reviewReason,
  };
  const decision = reviewReason
    ? { status: "review" as const, label: "사람 검수 필요", reason: reviewReason === "low_confidence" ? "필수 필드는 복구했지만 confidence가 90% 미만입니다." : "필수 필드가 모두 채워지지 않았습니다." }
    : { status: "eligible" as const, label: "자동 저장 가능", reason: "필수 필드 4개와 confidence 기준을 통과했습니다. 이 체험에서는 DB에 저장하지 않습니다." };
  return { product, evidence, required_field_coverage: coverage, dom_fingerprint: domFingerprint, warnings, decision };
}
